mlflow/java/client/src/main/java/org/mlflow/artifacts/ArtifactRepository.java

```diff
@@ -0,0 +1,102 @@
+package org.mlflow.artifacts;
+
+import java.io.File;
+import java.util.List;
+
+import org.mlflow.api.proto.Service.FileInfo;
+
+/**
+ * Allows logging, listing, and downloading artifacts against a remote Artifact Repository.
+ * This is used for storing potentially-large objects associated with MLflow runs.
+ */
+public interface ArtifactRepository {
+
+  /**
+   * Uploads the given local file to the run's root artifact directory. For example,
+   *
+   *  logArtifact("/my/localModel")
+   *  listArtifacts() // returns "localModel"
+   *
+   * @param localFile File to upload. Must exist, and must be a simple file (not a directory).
+   */
+  void logArtifact(File localFile);
+
+  /**
+   * Uploads the given local file to an artifactPath within the run's root directory. For example,
+   *
+   *   logArtifact("/my/localModel", "model")
+   *   listArtifacts("model") // returns "model/localModel"
+   *
+   * (i.e., the localModel file is now available in model/localModel).
+   *
+   * @param localFile File to upload. Must exist, and must be a simple file (not a directory).
+   * @param artifactPath Artifact path relative to the run's root directory. Should NOT
+   *                     start with a /. May be null.
```

Comment:
If we allow `null` it boils down to the previous API (or so it would seem). To avoid confusion, I recommend one of these:
- Only support `void logArtifact(File localFile, String artifactPath);` with a `@Nullable` annotation
- Have 2 APIs and the second one have an explicit `@Notnull` annotation

Reply:
We actually don't have a dependency which gives us `@Nullable` right now (except for com.sun.istack.internal). It's not very Java-y to invoke a method with null explicitly provided (so I would think a user would sort of expect non-null by default), which is why I exposed both methods. I think from a user's perspective, this is pretty normal, it only bloats the implementation somewhat.